pkg/kube/pod_controller.go

```diff
@@ -230,7 +230,66 @@ func (p *podController) StreamPodLogs(ctx context.Context) (io.ReadCloser, error
 		return nil, ErrPodControllerPodNotStarted
 	}
 
-	return StreamPodLogs(ctx, p.cli, p.pod.Namespace, p.pod.Name, ContainerNameFromPodOptsOrDefault(p.podOptions))
+	return newRestoreLogStreamReader(ctx, p.cli, p.pod.Namespace, p.pod.Name, ContainerNameFromPodOptsOrDefault(p.podOptions))
+}
+
+// newRestoreLogStreamReader creates a new restoreLogStreamReader instance which is used to stream logs from a pod.
+// restoreLogStreamReader will automatically try to establish a new log stream if the current one is closed and the pod is still alive.
+// This wrapper has to exist as there is hardcoded 4h timeout in kubelet for streaming logs.
+func newRestoreLogStreamReader(ctx context.Context, cli kubernetes.Interface, namespace, podName, containerName string) (io.ReadCloser, error) {
+	reader, err := StreamPodLogs(ctx, cli, namespace, podName, containerName, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &restoreLogStreamReader{
+		ctx:           ctx,
+		cli:           cli,
+		namespace:     namespace,
+		podName:       podName,
+		containerName: containerName,
+		reader:        reader,
+	}, nil
+}
+
+type restoreLogStreamReader struct {
+	ctx           context.Context
+	cli           kubernetes.Interface
+	namespace     string
+	podName       string
+	containerName string
+	reader        io.ReadCloser
+	lastReadTime  metav1.Time
+}
+
+func (s *restoreLogStreamReader) Read(p []byte) (n int, err error) {
+	n, err = s.reader.Read(p)
+	defer func() { s.lastReadTime = metav1.Now() }()
```

Comment:
This looks like we take the timestamp after the stream close. My main concern in https://github.com/kanisterio/kanister/pull/2903#discussion_r1616349780 was that if there was a log at the time of the stream close we might lose some records. 
Did you have a way to reproduce the failure so we can have at least empirical proof that this approach works?

Reply:
Yes, there is a way to reproduce it, and I did it. Unfortunately, each test takes 4 hours as I just run an action from a blueprint that uses a long `sleep` and `echo`.

This approach in manual tests always resulted in a duplicated last line of the log, which was independent of the frequency of the log. The last line was repeated even if it was the only line since 4 hours.

Reply:
Deduplication would be quite an overhead for what we are trying to achieve. This problem manifests only after 4 hours, and a single duplicated line for tasks that take over 4 hours, in my opinion, doesn't justify the need to add deduplication logic. This logic would require storing the last read line in memory and comparing it after reconnecting. Additionally, we should not assume that our users' logs distinguish between lines, which would complicate such a simple comparison.

Therefore, in my opinion, we should accept this single duplicated line that occurs on rare occasions.